Static compaction groups

raphaelsc opened the pull request:
Allows static configuration of number of compaction groups per table per shard.

To bootstrap the project, config option x_log2_compaction_groups was added which controls both number of groups and partitioning within a shard.

With a value of 0 (default), it means 1 compaction group, therefore all tokens go there.
With a value of 3, it means 8 compaction groups, and 3 most-significant-bits of tokens being used to decide which group owns the token.
And so on.

It's still missing:
- integration with repair / streaming
- integration with reshard / reshape.

perf/perf_simple_query --smp 1 --memory 1G

BEFORE
-----
median 61358.55 tps ( 71.1 allocs/op,  12.2 tasks/op,   56375 insns/op,        0 errors)
median 61322.80 tps ( 71.1 allocs/op,  12.2 tasks/op,   56391 insns/op,        0 errors)
median 61058.58 tps ( 71.1 allocs/op,  12.2 tasks/op,   56386 insns/op,        0 errors)
median 61040.94 tps ( 71.1 allocs/op,  12.2 tasks/op,   56381 insns/op,        0 errors)
median 61118.40 tps ( 71.1 allocs/op,  12.2 tasks/op,   56379 insns/op,        0 errors)

AFTER
-----
median 61656.12 tps ( 71.1 allocs/op,  12.2 tasks/op,   56486 insns/op,        0 errors)
median 61483.29 tps ( 71.1 allocs/op,  12.2 tasks/op,   56495 insns/op,        0 errors)
median 61638.05 tps ( 71.1 allocs/op,  12.2 tasks/op,   56494 insns/op,        0 errors)
median 61726.09 tps ( 71.1 allocs/op,  12.2 tasks/op,   56509 insns/op,        0 errors)
median 61537.55 tps ( 71.1 allocs/op,  12.2 tasks/op,   56491 insns/op,        0 errors)

bhalevy commented:
It's essential @tgrabiec reviews this too as it's highly relevant to tablets.

bhalevy commented on replica/database.hh:
```diff
@@ -409,7 +409,7 @@ private:
     // TODO: Still holds a single compaction group, meaning all sstables are eligible to be compacted with one another. Soon, a table
     //  will be able to hold more than one group.
-    std::unique_ptr<compaction_group> _compaction_group;
+    lw_shared_ptr<compaction_group> _compaction_group;
```
Side note, we need all async operations that depend on token mapping to hold on to the effective_replication_map shared instance they are based on.  Then, when there would be a concurrent operation that changes the mapping of tokens either to nodes or locally between compaction groups / tablets, I think we should use some kind of a barrier so to drain any operations that use the soon-to-be-deprecated e_r_m, and maybe consider aborting some of them that are not essential), then apply the updated e_r_m safely.
Some of the operations that were aborted could start over with the new e_r_m, e.g. cleanup compaction.

raphaelsc replied:
Isn't it important to allow cleanup completion to avoid data resurrection?

Ex:

- tablet t1 owns token [A, B]
- t1 is split. t1 owns A. t2 owns B. but both owns sstable storing [A, B]
- B is deleted on t2. tombstone is soon purged with shadowed data.
- cleanup starts running on t1 and t2.
- before cleanup finishes, tablet t1 and t2 are merged again into t1.
- t1 regains ownership of B, and ressurects B.

perhaps we could have versioned replication_map, which ongoing cleanup can use to discard data that's not supposed to be there. so we can have incremental (serialized) transformations on the data as topology changes. the new received data, as part of split, won't be touched by ongoing cleanup anyway, only the old data.

bhalevy replied:
yes, similar to node ops today where decommission and remove node need to wait until cleanup that followed bootstarp, tablet merging will need to block until cleanup following splis is done.

denesb replied:
We already have a mechanism for this: read and write phased barriers in the table object.

bhalevy commented on replica/table.cc:
```diff
@@ -493,6 +493,13 @@ compaction_group& table::compaction_group_for_sstable(const sstables::shared_sst
 }
 
+std::vector<lw_shared_ptr<compaction_group>> table::compaction_groups() const {
+    // FIXME: will return table::_compaction_groups, with no allocation required, once multiple groups are supported.
```
No allocation will be required if it will change to return `const std::vector<lw_shared_ptr<compaction_group>>&`

raphaelsc replied:
That's indeed my intention. I forgot to change the function signature in the end.

bhalevy commented on replica/database.hh:
```diff
@@ -710,9 +710,8 @@ public:
-    // as many active sstables as there are groups.
-    memtable& active_memtable();
+    // Return all active memtables, where there will be one per compaction group
+    std::vector<std::reference_wrapper<memtable>> active_memtables();
```
why don't we need a lw_shared_ptr<memtable> here like we do for compaction_group?

avikivity replied:
Why not memtable*?

raphaelsc replied:
Perhaps we can expose methods for operating the active memtables (retrieving statistics, etc) rather than exposing the memtables themselves.

avikivity replied:
std::reference_wrapper is equivalent to a pointer, just takes longer to write.

memtable lifetime management must be careful, more careful than lw_shared_ptr which only ensures the object is alive but says nothing about whether it should be alive in the first place.

raphaelsc replied:
> std::reference_wrapper is equivalent to a pointer, just takes longer to write.
> 
> memtable lifetime management must be careful, more careful than lw_shared_ptr which only ensures the object is alive but says nothing about whether it should be alive in the first place.

When I wrote that, the idea of using reference_wrapper was already abandoned in my mind. I think we need to change the interface. Instead of exposing memtable itself, expose what the current callers need from it.

bhalevy commented on replica/database.hh:
```diff
@@ -524,6 +524,7 @@ public:
                        const std::vector<sstables::shared_sstable>& old_sstables);
     };
 private:
+    using compaction_group_ptr = lw_shared_ptr<compaction_group>;
```
nit: this is not used in this patch

raphaelsc replied:
thanks

bhalevy commented on replica/table.cc:
```diff
@@ -292,7 +292,7 @@ future<std::vector<locked_cell>> table::lock_counter_cells(const mutation& m, db
 std::vector<std::reference_wrapper<memtable>> table::active_memtables() {
     return boost::copy_range<std::vector<std::reference_wrapper<memtable>>>(compaction_groups()
-            | boost::adaptors::transformed([] (lw_shared_ptr<compaction_group>& cg) {
+            | boost::adaptors::transformed([] (const lw_shared_ptr<compaction_group>& cg) {
```
why is that needed?

bhalevy commented on replica/table.cc:
```diff
@@ -147,6 +147,25 @@ table::find_row(schema_ptr s, reader_permit permit, const dht::decorated_key& pa
+        const tracing::trace_state_ptr& trace_state,
+        streamed_mutation::forwarding fwd,
+        mutation_reader::forwarding fwd_mr) const {
+    readers.reserve(readers.size() + _compaction_group->memtables()->size() + 1);
```
I don't think this reservation will help anything in the normal case as typically compaction groups will have zero to one memtables

bhalevy replied:
reserving too little might cause high write amplification of the record every time the re-reservation adds just enough extra space. See 7f80602b010a973dd24953cf39161699e7afcb42

avikivity commented:
> Allows static configuration of number of compaction groups per table per shard.
> 
> To bootstrap the project, config option x_log2_compaction_groups was added which controls both number of groups and partitioning within a shard.
> 
> With a value of 0 (default), it means 1 compaction group, therefore all tokens go there. With a value of 3, it means 8 compaction groups, and 3 most-significant-bits of tokens being used to decide which group owns the token. And so on.
> 
> It's still missing:
> 
> * integration with repair / streaming
> * integration with reshard / reshape.

We can go without these

> * automated testing

But let's have tests with the series.

avikivity commented on replica/database.hh:
```diff
@@ -409,7 +409,7 @@ private:
     // TODO: Still holds a single compaction group, meaning all sstables are eligible to be compacted with one another. Soon, a table
     //  will be able to hold more than one group.
-    std::unique_ptr<compaction_group> _compaction_group;
+    lw_shared_ptr<compaction_group> _compaction_group;
```
Can't we use regular pointers? I'd like to move away from the laziness of using shared pointers.

For static compaction groups, it shouldn't matter.

For dynamic compaction groups, we can use a barrier like @bhalevy suggests.

avikivity commented on replica/table.cc:
```diff
@@ -147,6 +147,25 @@ table::find_row(schema_ptr s, reader_permit permit, const dht::decorated_key& pa
+    for (auto&& mt : *_compaction_group->memtables()) {
+        if (auto reader_opt = mt->make_flat_reader_opt(s, permit, range, slice, pc, trace_state, fwd, fwd_mr)) {
+            readers.emplace_back(std::move(*reader_opt));
+        }
```
For point readers, we should only use one compaction group. For scans, we should open the readers one by one to avoid excessive merging.

avikivity replied:
I'm taking it partially back. Single-point reads are really easy and more important, so let's do them. We can wait with range reads.

However, please verify (via tracing) that a full scan still opens sstables incrementally. For example, reading the very first page should only create readers for sstables in the first compaction group.

avikivity replied:
I think with compound_sstable_set, it won't work. compound_sstable_set doesn't assume there is any ordering between the sstable sets, and always selects from all its inputs. So a scan from 32 compaction groups will merge from 32 at least readers.

If we had a tablet_sstable_set, it would select from tablets sequentially.

raphaelsc replied:
tablet_sstable_set may be really necessary for dynamic tablets, as once tablets are split, disjointness is broken at sstable level. so the set makes sure that only one compaction group will feed data only for the range it owns, avoiding read amplification too.

avikivity commented on replica/table.cc:
```diff
@@ -193,11 +211,7 @@ table::make_reader_v2(schema_ptr s,
-            readers.emplace_back(std::move(*reader_opt));
-        }
-    }
+    add_memtables_to_reader_list(readers, s, permit, range, slice, pc, trace_state, fwd, fwd_mr);
```
Here, we should either select one reader (single partition), or create a virtual reader that stitches together readers for relevant memtables (scans). The scan case should grab the memtable at the last possible moment, to avoid pinning it in memory.  

In fact the loop around compaction groups should enclose the cache readers (and create a new cache reader every time we cross a compaction group boundary).

avikivity replied:
Yes, I forgot the details here. A reader_selector should be enough.

btw, we'll have to get rid of it once we have tablets (also the multishard reader) since a tablet will map to just one shard. That means the role of the multishard reader will be assumed by storage_proxy::query_partition_key_range_concurrent.

raphaelsc replied:
@avikivity why can't we reuse the same cache reader? why do we have to create a new one when moving to next group? once we cross compaction group boundary, the cache reader can still feed the data for the current position.

similar to what happens when L0 sstables are read along with Ln sstables, where n > 0. L0 sstables span roughly the entire token range, meaning their readers will be alive most of the time.

combined reader
   incremental memtable reader selection (assumes disjointness)
   incremental sstable reader selection (assumes disjointness across groups)
   row cache reader

avikivity replied:
We can use the same reader, with an incremental selector.

The downside is that it pins sstables from all compaction groups, but it's not a regression from what we have now (where we pin all sstables anyway). No regression -> no requirement.

avikivity commented on replica/table.cc:
```diff
@@ -1633,12 +1633,12 @@ lw_shared_ptr<memtable_list>& compaction_group::memtables() noexcept {
-    });
+    auto fp = _highest_rp;
+    co_await _compaction_group->flush();
+    _flush_rp = std::max(_flush_rp, fp);
```
Please push this separately, it can be merged more quickly.

avikivity commented on replica/table.cc:
```diff
@@ -512,6 +512,13 @@ std::vector<lw_shared_ptr<compaction_group>> table::compaction_groups() const {
+    for (compaction_group_ptr& cg : cgs) {
+        co_await action(*cg);
+    }
+}
```
(here we can insert the barriers later once we have dynamic compaction groups)

raphaelsc replied:
I think I'll rename it to parallel_foreach_compaction_group(). Parallelism is really important as you want all groups to make progress on some requested table-wide operation, even if they're serialized down the road, like in compaction manager. For operations that don't need parallelism, we can go with a simple loop over replica::table::compaction_groups().

avikivity commented on replica/table.cc:
```diff
@@ -1651,7 +1651,9 @@ future<> table::flush(std::optional<db::replay_position> pos) {
 bool table::can_flush() const {
-    return _compaction_group->can_flush();
+    return std::ranges::all_of(compaction_groups(), [] (const compaction_group_ptr& cg) {
+        return cg->can_flush();
```
Similar, can use member pointer.

raphaelsc replied:
I don't think it was a bug, but rather partial support which resulted in compiler errors for unsupported features.

https://libcxx.llvm.org/Status/Cxx20.html

avikivity commented on replica/table.cc:
```diff
@@ -1212,13 +1212,18 @@ void table::set_compaction_strategy(sstables::compaction_strategy_type strategy)
 
+    for (compaction_group_ptr& cg : compaction_groups()) {
+        compaction_group_sstable_set_updater updater(*this, *cg, new_cs);
+        updater.prepare(new_cs);
```
Is it safe now to share a compaction strategy among multiple groups?

raphaelsc replied:
Not yet.

avikivity commented on replica/table.cc:
```diff
@@ -97,7 +97,12 @@ lw_shared_ptr<sstables::sstable_set> compaction_group::make_compound_sstable_set
+    }
+    auto sstable_sets = boost::copy_range<std::vector<lw_shared_ptr<sstables::sstable_set>>>(compaction_groups()
+        | boost::adaptors::transformed(std::mem_fn(&compaction_group::make_compound_sstable_set)));
+    return make_lw_shared(sstables::make_compound_sstable_set(schema(), std::move(sstable_sets)));
```
We should consider a custom solution here, since this pins all of the compaction groups' sstables, whereas we could pin just one. But this is no worse than what we have now, so fine for now.

raphaelsc replied:
This compound set spanning all groups is there for table operations which span all the groups. Point queries should pin a single sstable set. Range scans should still be able to incrementally read the sstables through the incremental selection procedure which compound set implements. Each compaction group implements its own incremental selector, and when reading from that compound set, the incremental selection still happens as expected as groups are disjoint in their token range.

avikivity replied:
Look at compound set selector, it merges the results from all selectors rather than sorting them.

avikivity replied:
It could use a heap to return them incrementally. But a tablet_sstable_set could just iterate based on the token.

raphaelsc replied:
you're right. I assumed partitioned incremental selection, but I can't do that. Also, L0 sstables are included unconditionally. We can teach compound set selector to work with one selector at a time, based on current position.

avikivity replied:
I'm guessing that a dedicated selector will be easier (but just a guess).

avikivity commented on replica/table.cc:
```diff
@@ -1736,22 +1736,20 @@ future<db::replay_position> table::discard_sstables(db_clock::time_point truncat
-            return make_ready_future<db::replay_position>(p->rp);
         });
-    });
+        co_return p->rp;
```
Please push separately.

avikivity commented on replica/table.cc:
```diff
@@ -1736,20 +1736,23 @@ future<db::replay_position> table::discard_sstables(db_clock::time_point truncat
-        auto p = make_lw_shared<pruner>(*this, *_compaction_group);
+    db::replay_position rp;
+    co_await for_each_compaction_group([this, truncated_at, &rp] (compaction_group& cg) -> future<> {
+        auto p = make_lw_shared<pruner>(*this, cg);
```
We should consider doing this in parallel to avoid delaying the commit log. Not critical.

avikivity commented on replica/table.cc:
```diff
@@ -1180,7 +1180,11 @@ future<bool> table::perform_offstrategy_compaction() {
-    return _compaction_manager.perform_offstrategy(as_table_state());
+    bool performed = false;
+    co_await for_each_compaction_group([this, &performed] (compaction_group& cg) -> future<> {
+        performed |= co_await _compaction_manager.perform_offstrategy(cg.as_table_state());
```
Should be done in parallel, so the compaction manager is informed of all the tasks and can cancel them. As written this cannot be cancelled.

avikivity commented on replica/table.cc:
```diff
@@ -1137,7 +1137,9 @@ compaction_group::update_main_sstable_list_on_compaction_completion(sstables::co
     co_await flush();
-    co_await _compaction_manager.perform_major_compaction(as_table_state());
+    co_await for_each_compaction_group([this] (compaction_group& cg) {
+        return _compaction_manager.perform_major_compaction(cg.as_table_state());
```
In parallel, compaction_manager will serialize as needed (and show the rest as pending).

avikivity commented on replica/table.cc:
```diff
@@ -577,7 +577,7 @@ table::update_cache(compaction_group& cg, lw_shared_ptr<memtable> m, std::vector
-        try_trigger_compaction();
+        try_trigger_compaction(cg);
     });
     if (cache_enabled()) {
```
Feels like this function should be moved to compaction_group.

Q: should we move row_cache to compaction_group too?

We don't need that to support tablets (we have invalidate_range) but maybe it makes things simpler.

avikivity replied:
It reduces efficiency too, so maybe not.

avikivity commented on replica/table.cc:
```diff
@@ -1149,22 +1149,23 @@ void table::start_compaction() {
-    do_trigger_compaction(); // see below
+    for (compaction_group_ptr& cg : compaction_groups()) {
+        do_trigger_compaction(*cg);
+    }
```
Should this be moved to compaction_group? For example, if an sstable is added to a compaction group, we only need to trigger compaction there.

raphaelsc replied:
Yes

avikivity commented on replica/database.cc:
```diff
@@ -2422,10 +2422,14 @@ future<> database::truncate_table_on_all_shards(sharded<database>& sharded_db, c
-            st->cres.emplace_back(co_await cm.stop_and_disable_compaction(cf.as_table_state()));
+            co_await cf.for_each_table_state([&cm, &st] (compaction::table_state& ts) -> future<> {
+                st->cres.emplace_back(co_await cm.stop_and_disable_compaction(ts));
+            });
```
Parallel?

avikivity commented on api/storage_service.cc:
```diff
@@ -754,7 +754,9 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
                 co_await run_on_existing_tables("upgrade_sstables", db, keyspace, table_infos, [&] (replica::table& t) {
-                    return t.get_compaction_manager().perform_sstable_upgrade(owned_ranges_ptr, t.as_table_state(), exclude_current_version);
+                    return t.for_each_table_state([&] (compaction::table_state& ts) {
+                        return t.get_compaction_manager().perform_sstable_upgrade(owned_ranges_ptr, ts, exclude_current_version);
```
Parallel?

raphaelsc replied:
Yes

avikivity commented on replica/table.cc:
```diff
@@ -489,12 +489,30 @@ void table::enable_off_strategy_trigger() {
 
+std::vector<lw_shared_ptr<compaction_group>> table::make_compaction_groups() {
+    std::vector<lw_shared_ptr<compaction_group>> ret;
+    auto number_of_cg = std::pow(2, _x_log2_compaction_groups);
```
1 << _log2_compaction_groups

raphaelsc commented on test/boost/sstable_test.hh:
```diff
@@ -55,8 +55,7 @@ public:
     auto try_flush_memtable_to_sstable(lw_shared_ptr<replica::memtable> mt) {
-        // FIXME: for multiple groups, rewrite this without directly referencing replica::table::_compaction_group.
-        return _cf->try_flush_memtable_to_sstable(*_cf->_compaction_group, mt, sstable_write_permit::unconditional());
+        return _cf->try_flush_memtable_to_sstable(*_cf->single_compaction_group_if_available(), mt, sstable_write_permit::unconditional());
```
IIRC, once I moved to lw_shared_ptr, the compiler no longer allowed me to access _compaction_group from this header.

raphaelsc commented on replica/database.hh:
```diff
@@ -405,11 +405,16 @@ private:
     compaction_manager& _compaction_manager;
     sstables::compaction_strategy _compaction_strategy;
     // TODO: Still holds a single compaction group, meaning all sstables are eligible to be compacted with one another. Soon, a table
     //  will be able to hold more than one group.
```
oops :-)

raphaelsc commented:
v2:
- Avoids lw_shared_ptr<compaction_group>
- table::compaction_groups() returns const ref
- table::active_memtables() returns std::vector<memtable*>
- don't reserve readers capacity in table::add_memtables_to_reader_list().
- introduce table::memtable_count() for memtable count across all groups
- point query will now pick memtables from a single compaction group
- use std::mem_fn() for reducing action on all groups to a single line, e.g.:
    std::ranges::any_of(compaction_groups(), std::mem_fn(&compaction_group::T));
- for_each_compaction_group(...) becomes parallel_foreach_compaction_group(...).
    we can operate on all groups in parallel
    if serialization is needed, that's done by lower layer like compaction_manager.
- table::do_trigger_compaction() is moved to compaction_group
- use 1 << _x_log2_compaction_groups when counting nr of groups.
- Reword description of option x_log2_compaction_groups
- Extend tests to stress multiple compaction groups

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3485/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3487/

avikivity commented on scylla-gdb.py:
```diff
@@ -4197,12 +4197,13 @@ def invoke(self, arg, from_tty):
+                for mt_ptr in std_vector(memtable_list['_memtables']):
+                    mt = seastar_lw_shared_ptr(mt_ptr).get()
+                    reg = lsa_region(mt.cast(region_ptr_type))
+                    gdb.write('  (memtable*) 0x%x: total=%d, used=%d, free=%d, flushed=%d\n' % (mt, reg.total(), reg.used(), reg.free(), mt['_flushed_memory']))
```
Please retain compatibility with older Scyllas. Although I see we already lost compatibility with pre-compaction-group scylla.

@denesb I don't think we can win this fight, since the tests don't care about older scylla.

denesb replied:
Yes, we usually try to enforce compatibility at review time and fix those that still sneak through later.

raphaelsc replied:
will fix.

avikivity commented on test/boost/commitlog_test.cc:
```diff
@@ -612,7 +612,7 @@ SEASTAR_TEST_CASE(test_commitlog_replay_invalid_key){
         auto& cl = *table.commitlog();
         auto s = table.schema();
-        auto& mt = table.active_memtable();
+        replica::memtable& mt = *table.active_memtables().front();
```
This works for a single compaction group, but won't it fail with multiple?

I'd like the test to pass with multiple compaction groups.

raphaelsc replied:
will try to adjust them for multiple groups, or override like you suggested.

avikivity commented on replica/table.cc:
```diff
@@ -157,10 +157,21 @@ table::add_memtables_to_reader_list(std::vector<flat_mutation_reader_v2>& reader
+    if (range.is_singular() && range.start()->value().has_key()) {
+        const dht::ring_position& pos = range.start()->value();
+        add_memtables_from_cg(compaction_group_for_token(pos.token()));
+    }
```
So the reserve() for the reader list is wrong in this case. We iterated on all compaction groups, but we should have picked just one.

raphaelsc replied:
There's also a bug here. After adding memtables from a single group, I should have returned.

avikivity commented on replica/table.cc:
```diff
@@ -527,6 +527,13 @@ const std::vector<std::unique_ptr<compaction_group>>& table::compaction_groups()
+future<> table::parallel_foreach_compaction_group(std::function<future<>(compaction_group&)> action) {
+    // TODO: place a barrier here when we allow dynamic groups.
+    co_await coroutine::parallel_for_each(compaction_groups(), [&] (const compaction_group_ptr& cg) {
+       return action(*cg);
```
Need 4 space indent, not 3.

avikivity replied:
Note: with tablets, cg.get() may be nullptr. But let's deal with it later.

raphaelsc replied:
oops, will fix indent.

raphaelsc replied:
> Note: with tablets, cg.get() may be nullptr. But let's deal with it later.

you're right.

avikivity commented on replica/table.cc:
```diff
@@ -1658,7 +1658,7 @@ future<> table::flush(std::optional<db::replay_position> pos) {
 
 bool table::can_flush() const {
-    return _compaction_group->can_flush();
+    return std::ranges::any_of(compaction_groups(), std::mem_fn(&compaction_group::can_flush));
```
What happens if table::can_flush() returns true for some compaction group, and then we call compaction_group::flush() even for compaction groups that can't?

raphaelsc replied:
While extending the tests, I changed table::flush() to issue compaction::flush() only if compaction_group::can_flush() is true.  I'll patch table::flush() again with that change.

avikivity commented on replica/table.cc:
```diff
@@ -332,7 +332,7 @@ api::timestamp_type compaction_group::min_memtable_timestamp() const {
 
 api::timestamp_type table::min_memtable_timestamp() const {
-    return _compaction_group->min_memtable_timestamp();
+    return *boost::range::min_element(compaction_groups() | boost::adaptors::transformed(std::mem_fn(&compaction_group::min_memtable_timestamp)));
```
Not sure this function is still used.

raphaelsc replied:
It's still used by tests, we can remove it later.

avikivity commented on replica/table.cc:
```diff
@@ -1738,7 +1738,11 @@ future<db::replay_position> table::discard_sstables(db_clock::time_point truncat
+
+    // FIXME: indentation
+    auto p = make_lw_shared<pruner>(*this, cg);
     co_await _cache.invalidate(row_cache::external_updater([p, truncated_at] {
```
The alternative is to invalidate the cache by token range (in fact it supports it already). Here is makes sense to iterate sequentually, since invalidation is CPU intensive work and doing it in parallel just competes with queries.

avikivity commented:
Looks good.

Please run the entire test suite (test.py) with multiple compaction groups.

Please adjust changelogs where you adapted a function with information about callers, and why the change is okay. For example, an function that is called from the api is fine to just accumulate, but rebuilding sstable sets after compaction is not okay to iterate over all compaction groups.

raphaelsc commented on replica/table.cc:
```diff
@@ -1653,7 +1653,7 @@ future<> table::flush(std::optional<db::replay_position> pos) {
     auto op = _pending_flushes_phaser.start();
     auto fp = _highest_rp;
-    co_await _compaction_group->flush();
+    co_await parallel_foreach_compaction_group(std::mem_fn(&compaction_group::flush));
```
I'll add a comment, so we can revisit it later.

raphaelsc commented:
> Looks good.
> 
> Please run the entire test suite (test.py) with multiple compaction groups.

Some unit tests were written with a single compaction group in mind and they fail if I force they to run with multiple groups, e.g. they expect a certain # of sstables per shard, not one per group. I am thinking about incrementally enabling multiple groups until the entire suite is ready for it. database_test and mutation_test will already be stressing multiple groups, even with default settings in test.py. We can extend that to other relevant boost tests. For tests which launch a scylla instance, those can be easily forced to run with multiple groups with a simple patch to test.py.

> 
> Please adjust changelogs where you adapted a function with information about callers, and why the change is okay. For example, an function that is called from the api is fine to just accumulate, but rebuilding sstable sets after compaction is not okay to iterate over all compaction groups.

raphaelsc commented:
> > Looks good.
> > Please run the entire test suite (test.py) with multiple compaction groups.
> 
> Some unit tests were written with a single compaction group in mind and they fail if I force they to run with multiple groups, e.g. they expect a certain # of sstables per shard, not one per group. I am thinking about incrementally enabling multiple groups until the entire suite is ready for it. database_test and mutation_test will already be stressing multiple groups, even with default settings in test.py. We can extend that to other relevant boost tests. For tests which launch a scylla instance, those can be easily forced to run with multiple groups with a simple patch to test.py (via a new option).
> 
> > Please adjust changelogs where you adapted a function with information about callers, and why the change is okay. For example, an function that is called from the api is fine to just accumulate, but rebuilding sstable sets after compaction is not okay to iterate over all compaction groups.

Good news is that only 8 out of 2607 tests failed, with multiple groups.

raphaelsc commented:
v3:
Fixed backward compatibility in scylla-memtables gdb command.
Adjusted changelogs for table methods adapted to multiple groups
Extended pruner to operate on all groups, in order to invalidate cache only once
Fixed indentation in parallel_foreach_compaction_group()
Added missing FIXME's
Fixed reserve() when adding memtables of compaction groups

BEFORE
-----
median 61358.55 tps ( 71.1 allocs/op,  12.2 tasks/op,   56375 insns/op,        0 errors)
median 61322.80 tps ( 71.1 allocs/op,  12.2 tasks/op,   56391 insns/op,        0 errors)
median 61058.58 tps ( 71.1 allocs/op,  12.2 tasks/op,   56386 insns/op,        0 errors)
median 61040.94 tps ( 71.1 allocs/op,  12.2 tasks/op,   56381 insns/op,        0 errors)
median 61118.40 tps ( 71.1 allocs/op,  12.2 tasks/op,   56379 insns/op,        0 errors)

AFTER
-----
median 61656.12 tps ( 71.1 allocs/op,  12.2 tasks/op,   56486 insns/op,        0 errors)
median 61483.29 tps ( 71.1 allocs/op,  12.2 tasks/op,   56495 insns/op,        0 errors)
median 61638.05 tps ( 71.1 allocs/op,  12.2 tasks/op,   56494 insns/op,        0 errors)
median 61726.09 tps ( 71.1 allocs/op,  12.2 tasks/op,   56509 insns/op,        0 errors)
median 61537.55 tps ( 71.1 allocs/op,  12.2 tasks/op,   56491 insns/op,        0 errors)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3526/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3549/

raphaelsc commented:
v4: rebased

avikivity commented:
So, the test suite now passes with multiple compaction groups (not talking about just the dedicated tests)?

raphaelsc commented:
> So, the test suite now passes with multiple compaction groups (not talking about just the dedicated tests)?

@avikivity 7 tests out 2637 failed. most of them fail because they were written with a single group in mind, like expecting a certain # of sstables. I plan to address the failure after this series is merged. By the way, we need to run the suite with multiple groups on a regular base, I am planning to add a compilation option to override the default # of groups. I can also patch test.py to optionally launch scylla instances with multiple groups, but that won't cover boost and similar tests. For boost, we're running both single and multiple groups on a subset of tests and we can keep extending this subset as we wish.

avikivity commented:
ok.
